rfd/0194-k8s-rbac-simplification.md

```diff
@@ -0,0 +1,209 @@
+- Clarify / Improve documentation for the various ways to setup Kubernetes in
+  Teleport.
+
+## Proposal
+
+### Background
+
+Currently, when enrolling a Kubernetes cluster and setting it up to use the
+group `system:masters`, everything works as expected (except on GKE Autopilot
+where this group is forbidden).
+Teleport is super admin and yields reduced permissions to the user based on
+the `kubernetes_resources` and `kubernetes_label` fields from the Teleport
+role.
+
+### Impersonation
+
+To limit the number of code changes, the proposal is to default the
+impersonation to use the Teleport service account.
+
+We will need to make sure the service account has `cluster-admin` permissions.
+
+To avoid unexpected access being granted, the `kubernetes_resources` field
+should default to `null` as well instead of the current wildcard.
+
+The `kubernetes_user` should default to the Teleport user auth email.
+
+### UX
+
+#### Deprecate `namespace`
+
+We currently have the `namespace` kind mean "everything within a namespace",
+which has caused confusion with customers and resulting in unexpected access
+granted.
+
+This should be deprecated and removed. It would require a bump on the role
+model to v8 to avoid changing behavior for current customers.
+
+Upgrading to v8 would map the deprecated `namespace` field to `*` while
+swapping `name`/`namespace` as these are equivalent within a namespace:
+
+```yaml
+kind: namespace
+name: foo
+```
+
+and:
+
+```yaml
+kind: '*'
+namespace: foo
+```
+
+Note that the later also grants access to cluster-wide resources outside a
+namesace.
+
+To avoid confustion, this behavior would change and when a `namespace` value is
+set (unless it is `*`), only namespaced resources would be matched.
+To match cluster-wide resource, the namespace field should be omitted or set to
+`*`.
+
+#### CRD / Subresources support
+
+Currently, we only support a hard-coded list of pre-set resources. To enable
+CRD support, i.e. arbitrary kind, a new field `api` will be introduced.
+When not set, the `kind` field would still be restricted to the existing
+hard-coded list, but once set, it can be an arbitrary string.
+
+Similirally, to avoid confusion around sub-resources and the common mistake
+of granting `exec` permission with the `get` verb, a new field `sub_resources`
+will be introduce. If not set, then only the main resource would match, which
+removes the unexpected side-effect for `exec`.
+To avoid confusion, when `sub_resources` are set, the `verbs` field will be
+ignored.
+
+The role would look like this:
+
+```yaml
+kubernetes_resources:
+  - kind: '*'
+    api: 'API with version'
+    name: '*'
+    namespace: '*'
+    verbs: ['*']
+    sub_resources: ['*']
+```
+
+#### Web Role Editor
+
+To avoid confusion, in the web role editor, we would remove the ability to set
+and view `kubernetes_group` and `kubernetes_users` in favor of only allowing it
+to be controlled in yaml.
```

Comment:
The preset `require-trusted-device` role sets `kubernetes_group` field. Currently we don't support it fully, because of a missing option, but if we wanted to fully support it, removing this field from the editor would prevent us from doing so.